tool/tsh/common/app.go

```diff
@@ -52,26 +52,65 @@ func onAppLogin(cf *CLIConf) error {
 	if err != nil {
 		return trace.Wrap(err)
 	}
+
+	profile, err := tc.ProfileStatus()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
 	app, err := getRegisteredApp(cf, tc)
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	profile, err := tc.ProfileStatus()
+
+	routeToApp, err := getRouteToApp(cf, tc, profile, app)
```

Comment:
Note to reviewers: the refactors in this file should have no functional implications - I just needed to reuse some of this logic. Some split-out methods were only needed for early iterations of this change, but it's easier (better?) not to revert these changes now.